dali/imgcodec/parsers/bmp.cc

```diff
@@ -12,19 +12,150 @@
+};
+
+bool is_color_palette(InputStream *stream, int ncolors, int palette_entry_size) {
+  SmallVector<uint8_t, 8> entry;
+  entry.resize(palette_entry_size);
+  for (int i = 0; i < ncolors; i++) {
+    stream->ReadAll(entry.data(), palette_entry_size);
+
+    const auto b = entry[0], g = entry[1], r = entry[2];  // a = p[3];
+    if (b != g || b != r)
+      return true;
+  }
+  return false;
+}
+
+int number_of_channels(InputStream *stream,
+                       int bpp, int compression_type, int ncolors = 0,
+                       int palette_entry_size = 0) {
+  if (compression_type == BMP_COMPRESSION_RGB || compression_type == BMP_COMPRESSION_RLE8) {
+    if (bpp <= 8 && ncolors <= (1_i64 << bpp)) {
+      return is_color_palette(stream, ncolors, palette_entry_size) ? 3 : 1;
+    } else if (bpp == 24) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  } else if (compression_type == BMP_COMPRESSION_BITFIELDS) {
+    if (bpp == 16) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  }
+
+  throw std::runtime_error(make_string(
+    "configuration not supported. bpp: ", bpp, " compression_type:", compression_type,
+    "ncolors:", ncolors));
+}
+
+}  // namespace
+
+struct BitmapCoreHeader {
+  uint32_t header_size;
+  uint16_t width, heigth, planes, bpp;
+};
+static_assert(sizeof(BitmapCoreHeader) == 12);
+
+struct BitmapInfoHeader {
+  int32_t header_size;
+  int32_t width, heigth;
+  uint16_t planes, bpp;
+  uint32_t compression, image_size;
+  int32_t x_pixels_per_meter, y_pixels_per_meter;
+  uint32_t colors_used, colors_important;
+};
+static_assert(sizeof(BitmapInfoHeader) == 40);
+
 ImageInfo BmpParser::GetInfo(ImageSource *encoded) const {
+  auto stream = encoded->Open();
+  ssize_t length = stream->Size();
+
+  DALI_ENFORCE(length >= 18);
+
+  static constexpr int kHeaderStart = 14;
+  stream->SeekRead(kHeaderStart);
+  uint32_t header_size = stream->ReadOne<uint32_t>();
+  stream->Skip<uint32_t>(-1);  // we'll read it again - it's part of the header struct
+  int64_t h = 0, w = 0, c = 0;
+  int bpp = 0, compression_type = BMP_COMPRESSION_RGB;
+  int ncolors = 0;
+  int palette_entry_size = 0;
+  ptrdiff_t palette_start = 0;
+
+  if (length >= 26 && header_size == 12) {
+    BitmapCoreHeader header = {};
+    stream->ReadAll(&header, 1);
+    w = header.width;
+    h = header.heigth;
+    bpp = header.bpp;
+    if (bpp <= 8) {
+      palette_start = stream->TellRead();
+      palette_entry_size = 3;
+      ncolors = (1_uz << bpp);
+    }
+  } else if (length >= 50 && header_size >= 40) {
+    BitmapInfoHeader header = {};
+    stream->ReadAll(&header, 1);
+    w = abs(header.width);
+    h = abs(header.heigth);
+    bpp = header.bpp;
+    compression_type = header.compression;
+    ncolors = header.colors_used;
+    if (bpp <= 8) {
+      palette_start = stream->TellRead();
+      palette_entry_size = 4;
+      ncolors = ncolors == 0 ? 1_uz << bpp : ncolors;
+    }
+  }
```

Comment:
What if none of the two branches is taken? Are fine with calling number_of_channels (I guess) and returning zeros?

Reply:
Looking at the info here: https://en.wikipedia.org/wiki/BMP_file_format#DIB_header_(bitmap_information_header)
I understand the following:
- We support BITMAPCOREHEADER (12 bytes long)
- We support BITMAPINFOHEADER and all following versions (40, 52, 56, 108, 124 bytes long respectively)
Given that, we should not fall outside of those two cases. If that happens, the arguments of number_of_channels will be 0 which will result in an exception thrown.

@mzient can you double check my understanding is correct?

Reply:
If none of the two branches is taken then the file is malformed and parsing should fail - preferably here and not somewhere later.

Reply:
I've added an error.